test/functional/km/control.cpp

```diff
@@ -0,0 +1,489 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    Kernel Mode Test Driver
+
+--*/
+
+#include <ntddk.h>
+#include <wdf.h>
+#include <ntstrsafe.h>
+#include <netioddk.h>
+#include <ws2def.h>
+#include "cxplat.h"
+
+#include <xdpapi.h>
+
+#include "tests.h"
+#include "xdpfunctionaltestdrvioctl.h"
+#include "fntrace.h"
+
+#include "control.tmh"
+
+
+DECLARE_CONST_UNICODE_STRING(TestDrvCtlDeviceName, L"\\Device\\" FUNCTIONAL_TEST_DRIVER_NAME);
+DECLARE_CONST_UNICODE_STRING(TestDrvCtlDeviceSymLink, L"\\DosDevices\\" FUNCTIONAL_TEST_DRIVER_NAME);
+
+typedef struct TEST_CLIENT
+{
+    bool TestFailure;
+
+} TEST_CLIENT;
+
+WDF_DECLARE_CONTEXT_TYPE_WITH_NAME(TEST_CLIENT, TestDrvCtlGetFileContext);
+
+EVT_WDF_IO_QUEUE_IO_DEVICE_CONTROL TestDrvCtlEvtIoDeviceControl;
+EVT_WDF_IO_QUEUE_IO_CANCELED_ON_QUEUE TestDrvCtlEvtIoCanceled;
+
+PAGEDX EVT_WDF_DEVICE_FILE_CREATE TestDrvCtlEvtFileCreate;
+PAGEDX EVT_WDF_FILE_CLOSE TestDrvCtlEvtFileClose;
+PAGEDX EVT_WDF_FILE_CLEANUP TestDrvCtlEvtFileCleanup;
+
+WDFDEVICE TestDrvCtlDevice = nullptr;
+TEST_CLIENT* TestDrvClient = nullptr;
+CHAR TestDrvClientActive = FALSE;
+
+_No_competing_thread_
+INITCODE
+NTSTATUS
+TestDrvCtlInitialize(
+    _In_ WDFDRIVER Driver
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    PWDFDEVICE_INIT DeviceInit = nullptr;
+    WDF_FILEOBJECT_CONFIG FileConfig;
+    WDF_OBJECT_ATTRIBUTES Attribs;
+    WDFDEVICE Device;
+    WDF_IO_QUEUE_CONFIG QueueConfig;
+    WDFQUEUE Queue;
+
+    DeviceInit =
+        WdfControlDeviceInitAllocate(
+            Driver,
+            &SDDL_DEVOBJ_SYS_ALL_ADM_ALL);
+    if (DeviceInit == nullptr) {
+        TraceError(
+            "[ lib] ERROR, %s.",
```

Comment:
can we use a more descriptive prefix than `[ lib]`? maybe `[fnkt]` or similar?

Reply:
@nigriMSFT what do you prefer?

Reply:
Yes please improve this, [fnkt] seems fine. Lots of this was just copied from msquic, so some of the naming probably doesn't make sense here in the xdp repo